### PR TITLE
langchain_community.vectorstores: PGVector filtering - top level AND and OR operator #17064

### PR DESCRIPTION
**Pull request by djo10**
Solution for #17064 
**Description:**
- Included top level AND and OR operators for filtering eg: `filter = {"or": [{"rating": {"gte": 8.5}}, {"genre": "animated"}]}`
- included GTE and LTE comparators
- Code optimized and refactored, using Comparator and Operator classes from `libs/langchain/chains/query_constructor/ir.py`

### CODE REVIEW

**Comment by djo10** [comment]
@eyurtsev

File: libs/community/langchain_community/vectorstores/pgvector.py [hunk 1]
```diff
@@ -508,65 +508,61 @@ def _results_to_docs_and_scores(self, results: Any) -> List[Tuple[Document, floa
...
-                self._create_filter_clause(key, sub_value)
-                for sub_value in value_case_insensitive[AND]
-            ]
-            filter_by_metadata = sqlalchemy.and_(*and_clauses)
-
-        else:
-            filter_by_metadata = None
-
-        return filter_by_metadata
+    def _create_filter_clause(self, filters):  # type: ignore[no-untyped-def]
```
Comment (eyurtsev) [inline libs/community/langchain_community/vectorstores/pgvector.py:511]: Could we add a type for the filters?

File: libs/community/langchain_community/vectorstores/pgvector.py [hunk 2]
```diff
@@ -508,65 +508,61 @@ def _results_to_docs_and_scores(self, results: Any) -> List[Tuple[Document, floa
...
+            Comparator.GTE: lambda f, v: f >= str(v),
+            Comparator.LIKE: lambda f, v: f.like(v),
+            Comparator.IN: lambda f, v: f.in_(v),
+            Comparator.NIN: lambda f, v: f.not_in(v),
+            Comparator.CONTAIN: lambda f, v: f.contains(v),
+            "contains": lambda f, v: f.contains(v),
+            "between": lambda f, v: f.between(str(v[0]), str(v[1])),
+        }
+
+        def parse_condition(field, condition):  # type: ignore[no-untyped-def]
```
Comment (eyurtsev) [inline libs/community/langchain_community/vectorstores/pgvector.py:529]: Is it possible to add type annotations here to make it easier to follow what's going on?

File: libs/community/langchain_community/vectorstores/pgvector.py [hunk 3]
```diff
@@ -508,65 +508,61 @@ def _results_to_docs_and_scores(self, results: Any) -> List[Tuple[Document, floa
...
-
-        else:
-            filter_by_metadata = None
-
-        return filter_by_metadata
+    def _create_filter_clause(self, filters):  # type: ignore[no-untyped-def]
+        from langchain.chains.query_constructor.ir import Comparator, Operator
+
+        COMPARISONS = {
+            Comparator.EQ: lambda f, v: f == str(v),
```
Comment (eyurtsev) [inline libs/community/langchain_community/vectorstores/pgvector.py:515]: What happens if value was an integer and the field is of an integer type?

File: libs/community/langchain_community/vectorstores/pgvector.py [hunk 4]
```diff
@@ -508,65 +508,61 @@ def _results_to_docs_and_scores(self, results: Any) -> List[Tuple[Document, floa
...
+            Comparator.NE: lambda f, v: f != str(v),
+            Comparator.LT: lambda f, v: f < str(v),
+            Comparator.LTE: lambda f, v: f <= str(v),
+            Comparator.GT: lambda f, v: f > str(v),
+            Comparator.GTE: lambda f, v: f >= str(v),
+            Comparator.LIKE: lambda f, v: f.like(v),
+            Comparator.IN: lambda f, v: f.in_(v),
+            Comparator.NIN: lambda f, v: f.not_in(v),
+            Comparator.CONTAIN: lambda f, v: f.contains(v),
+            "contains": lambda f, v: f.contains(v),
```
Comment (eyurtsev) [inline libs/community/langchain_community/vectorstores/pgvector.py:525]: cc @baskaryan looks like we've got some operations not covered by the existing enums?

**Comment by eyurtsev** [comment]
Awesome! I'm wondering are there any existing unit tests that help cover the internal refactor? Could you confirm you ran the unit tests? The code looks good from a high level, but hard for me to determine whether it works without digging in.

**Comment by 765144989** [comment]
filter = {
    "OR": [
        {"age": {"GT": 30}},
        {"salary": {"EQ": 1000}}
    ]
}
dont work why??

**Comment by eyurtsev** [comment]
closing in favor of: https://github.com/langchain-ai/langchain/pull/18992

**Pull request closed without merging**